src/edu/washington/escience/myria/operator/CSVFileScanFragment.java

```diff
@@ -200,99 +206,76 @@ protected TupleBatch fetchNextReady() throws IOException, DbException {
-                  || Math.min(movingEndByte, maxByteRange) == maxByteRange) {
-                newLineFound = true;
-                // Reading from the source from the beginning again since we only have the character offset information
-                InputStream completePartitionStream =
-                    source.getInputStream(
-                        adjustedStartByteRange, Math.min(movingEndByte, maxByteRange));
-                BufferedReader reader =
-                    new BufferedReader(new InputStreamReader(completePartitionStream));
-                reader.skip(characterPositionAtBeginningOfRecord);
-                parser =
-                    new CSVParser(
-                        reader,
-                        CSVFormat.newFormat(delimiter).withQuote(quote).withEscape(escape),
-                        0,
-                        0);
-                iterator = parser.iterator();
-                record = iterator.next();
-                finishedReadingLastRow = true;
-                break;
-              }
-              dataChar = partitionInputStream.read();
+            if (finalLineFound) {
+              long characterPositionAtBeginningOfRecord = record.getCharacterPosition();
+              InputStream completePartitionStream =
+                  source.getInputStream(
+                      adjustedStartByteRange, Math.min(trailingEndByte, maxByteRange));
+              BufferedReader reader =
+                  new BufferedReader(new InputStreamReader(completePartitionStream));
+              reader.skip(characterPositionAtBeginningOfRecord);
+              parser =
+                  new CSVParser(
+                      reader,
+                      CSVFormat.newFormat(delimiter).withQuote(quote).withEscape(escape),
+                      0,
+                      0);
+              iterator = parser.iterator();
+              record = iterator.next();
+              finishedReadingLastRow = true;
+            } else {
+              byteOverlap *= 2;
+              trailingStartByte += byteOverlap;
+              trailingEndByte += byteOverlap;
             }
-            partitionInputStream.close();
-            byteOverlap *= 2;
           }
         } else if (record.size() == schema.numColumns() && onLastRow && isLastWorker) {
-          /*
-           * If we are on the last worker, just mark it as finished.
-           */
           finishedReadingLastRow = true;
         }
 
         /*
-         * At this point, we've exhausted all other cases and check for two conditions. 1) The record is not discarded
-         * and 2) If the we are on the last row, we need to make sure we've finished reading the entire row
+         * If we're on the last row, we check if we've finished reading the row completely.
          */
-        if (((!onLastRow) || (onLastRow && finishedReadingLastRow)) && !discardedRecord) {
+        if (((!onLastRow) || (onLastRow && finishedReadingLastRow))) {
```

Comment:
Nitpick: removing parens around `!onLastRow` makes it more readable to me (but maybe not you).